addons/nuxeo-template-rendering/nuxeo-template-rendering-core/src/main/java/org/nuxeo/template/processors/AbstractBindingResolver.java

```diff
@@ -70,106 +78,179 @@ protected DocumentObjectWrapper getWrapper() {
+            } catch (NoValueToAddInContext e) {
+                log.debug("Skip param to add " + param.getName(), e);
+            }
+        }
+    }
+
+    protected Object extractValueFromParam(TemplateBasedDocument templateBasedDocument, TemplateInput param) throws ValueNotFound, NoValueToAddInContext {
+        DocumentModel doc = templateBasedDocument.getAdaptedDoc();
+        String propKey = param.getSource();
+
+        switch (param.getType()) {
+            case BooleanValue:
+                return param.getBooleanValue();
+            case DateValue:
+                return param.getDateValue();
+            case StringValue:
+                return param.getStringValue();
+            case MapValue:
+                Map<String, Object> resultMap = new HashMap<>();
+                Map<String, TemplateInput> map = param.getMapValue();
+                for (String paramName : map.keySet()) {
+                    try {
+                        resultMap.put(paramName, extractValueFromParam(templateBasedDocument, map.get(paramName)));
+                    } catch (NoValueToAddInContext | ValueNotFound e) {
+                        log.debug("Skip param to add " + paramName + " in " + param.getName(), e);
                     }
-                    Property property = null;
+                }
+                return resultMap;
+            case ListValue:
+                List<Object> resultList = new ArrayList<>();
+                for (TemplateInput childParam : param.getListValue()) {
                     try {
-                        property = templateBasedDocument.getAdaptedDoc().getProperty(param.getSource());
-                    } catch (PropertyException e) {
-                        log.warn("Unable to ready property " + param.getSource(), e);
+                        resultList.add(extractValueFromParam(templateBasedDocument, childParam));
+                    } catch (NoValueToAddInContext | ValueNotFound e) {
+                        log.debug("Skip param to add " + childParam.getName() + " in " + param.getName(), e);
                     }
-
-                    Serializable value = null;
-                    if (property != null) {
-                        value = property.getValue();
+                }
+                return resultList;
+            case Content:
+                ContentInputType contentInput = ContentInputType.getByValue(param.getSource());
+                if (BlobContent.equals(contentInput))
+                    return extractBlobContent(doc, param);
+                else if (HtmlPreview.equals(contentInput))
+                    return extractHTMLPreview(doc, param);
+                else {
+                    Serializable docPropertyValue = getDocPropertyValue(doc, propKey);
+                    if (docPropertyValue instanceof String)
+                        return handleHtmlField(param.getName(), (String) getDocPropertyValue(doc, propKey));
+                }
```

Comment:
You're probably missing a `break` here.

Reply:
@efge : We don't really need that as we always return something in it => if / if else / else
The last else return something... Or I can break and throw exception in the bottom of the method (that should never happens)... Tell me